### PR TITLE
shared utility function to check if a cluster is workload identity

### PR DESCRIPTION
**Pull request by slawande2**
### Which issue this PR addresses:
[ARO-7856](https://issues.redhat.com/browse/ARO-7856)

### What this PR does / why we need it:
Throughout the RP codebase and installer wrapper, we are using if platformworkloadidentityprofile != nil or ServicePrincipalProfile to indicate whether or not a cluster is workload identity or not. This PR includes a shared utility function to check if a function is a Workload Identity or Service Principal cluster, which can be imported in all parts of the codebase.

### Test plan for issue:
Passed all the unit test cases.

### CODE REVIEW

**Review by mociarain — DISMISSED** [review]
Nice

File: pkg/util/iswimi/iswimi.go
Comment (bitoku) [inline pkg/util/iswimi/iswimi.go:1]: How about just `wimi` ?
`iswimi` is too specific and I'd hesitate to add new functions in this util package.

File: pkg/util/iswimi/iswimi.go
Comment (bitoku) [inline pkg/util/iswimi/iswimi.go:9]: nit: I would make this argument `OpenShiftCluster`, not `OpenShiftClusterProperties`.
It would make a bit more sense to me and it accords with the argument name.

File: pkg/util/iswimi/iswimi.go
Comment (kimorris27) [inline pkg/util/iswimi/iswimi.go:8]:
```suggestion
// IsWimi checks whether a cluster is a workload identity cluster or a service principal cluster
```

I'm suggesting this change because "classic" refers to a non-HCP cluster as opposed to an HCP cluster. We will eventually have both _WIMI classic clusters_ and _WIMI HCP clusters_.

File: pkg/util/iswimi/iswimi_test.go
Comment (kimorris27) [inline pkg/util/iswimi/iswimi_test.go:28]:
```suggestion
			name: "Cluster is Service Principal",
```

I'm suggesting this change for the same reason as explained in my doc comment on the `isWimi()` function.

**Review by kimorris27 — CHANGES_REQUESTED** [review]
LGTM broadly speaking! I second Ayato's suggestions, and I have one other small change request to make sure some terminology usage is clear.

File: pkg/util/iswimi/iswimi.go
Comment (cadenmarchese) [reply pkg/util/iswimi/iswimi.go:9]: Super nit - I think we should also use `oc` as the param name instead of `cluster` since it's usually what we use.

File: pkg/util/iswimi/iswimi.go
Comment (cadenmarchese) [inline pkg/util/iswimi/iswimi.go:10]: For example:
```suggestion
	if oc.PlatformWorkloadIdentityProfile == nil || oc.ServicePrincipalProfile != nil {
```

**Review by cadenmarchese — COMMENTED** [review]
great work here!

**Comment by slawande2** [comment]
The iswimi.go and iswimi_test.go files got deleted and recreated because I changed the package name

File: pkg/frontend/openshiftcluster_putorpatch.go
Comment (kimorris27) [inline pkg/frontend/openshiftcluster_putorpatch.go:139]: Nit: This line with the `TODO` comment can be removed since this PR implements the change it describes.

**Review by kimorris27 — COMMENTED** [review]
Thanks for being flexible in the review process! LGTM as long as E2E passes.

File: pkg/util/wimi/iswimi.go
Comment (bitoku) [inline pkg/util/wimi/iswimi.go:14]:
```suggestion
	if oc.Properties.PlatformWorkloadIdentityProfile != nil && oc.Properties.ServicePrincipalProfile == nil {
		return true
	}
	return false
}
```

If we get invalid cluster (e.g.  PlatformWorkloadIdentityProfile != nil and ServicePrincipalProfile != nil), it returns true, which is not correct.

We might not have to care about invalid settings, but it would be more intuitive that only when it has valid settings and WI is not nil, it returns true.

**Review by SudoBrendan — DISMISSED** [review]
LGTM - feel free to action or dismiss my nit.

File: pkg/util/iswimi/iswimi.go
Comment (slawande2) [reply pkg/util/iswimi/iswimi.go]: done, thank you!

File: pkg/operator/deploy/deploy.go
Comment (slawande2) [reply pkg/operator/deploy/deploy.go]: done, thank you! :)

File: pkg/util/wimi/iswimi.go
Comment (slawande2) [reply pkg/util/wimi/iswimi.go]: done, thank you!

File: pkg/cluster/delete.go
Comment (slawande2) [reply pkg/cluster/delete.go]: done, thank you!

File: pkg/util/wimi/iswimi_test.go
Comment (slawande2) [reply pkg/util/wimi/iswimi_test.go]: done, thank you!

File: pkg/frontend/openshiftcluster_putorpatch.go
Comment (slawande2) [reply pkg/frontend/openshiftcluster_putorpatch.go]: done, thank you!

File: pkg/frontend/openshiftcluster_putorpatch.go
Comment (slawande2) [reply pkg/frontend/openshiftcluster_putorpatch.go]: done, thank you!

File: pkg/cluster/deploybaseresources.go
Comment (slawande2) [reply pkg/cluster/deploybaseresources.go]: done, thank you!

File: pkg/util/iswimi/iswimi.go
Comment (slawande2) [reply pkg/util/iswimi/iswimi.go]: done, thank you!

**Review by kimorris27 — APPROVED** [review]
Nice work here.

I hadn't thought about it during my own reviews, but I think Ayato's suggestion [here](https://github.com/Azure/ARO-RP/pull/3683#discussion_r1680784794) makes sense if you feel inclined to make it before we merge. I don't consider it blocking though.

File: pkg/util/wimi/iswimi.go
Comment (slawande2) [reply pkg/util/wimi/iswimi.go]: done, thank you!

**Review by bitoku — APPROVED** [review]

**Review by kimorris27 — APPROVED** [review]

**Review by cadenmarchese — APPROVED** [review]

File: pkg/util/wimi/iswimi_test.go
Comment (rajdeepc2792) [inline pkg/util/wimi/iswimi_test.go:13]: Suuuper NIT:- Since the function is named `isWimi`, this can be changed to `TestIsWimi`.
Please feel free to dismiss it or handle it accordingly. ```suggestion
func TestIsWimi(t *testing.T) {
```

File: pkg/cluster/delete.go
Comment (AldoFusterTurpin) [inline pkg/cluster/delete.go:447]: Just a small comment: If we move the definition of the IsWimi function to the OpenShift cluster, we obtain a richer domain where functions are closer where the data is used (to avoid having an anemic Domain Model, a concept of DDD that here could be interesting). 
If doing that (check next comments) that line would look like:
```go
if m.doc.OpenShiftCluster.IsWimi() {
```
which is pretty neat as clearly states that the fact that a cluster is Wimi or not just depend on the cluster and reads a bit better 🙂

File: pkg/cluster/deploybaseresources.go
Comment (AldoFusterTurpin) [inline pkg/cluster/deploybaseresources.go:43]: Same as before.

File: pkg/frontend/openshiftcluster_putorpatch.go
Comment (AldoFusterTurpin) [inline pkg/frontend/openshiftcluster_putorpatch.go:139]: Same as before.

File: pkg/operator/deploy/deploy.go
Comment (AldoFusterTurpin) [inline pkg/operator/deploy/deploy.go:472]: Same as before.

File: pkg/util/wimi/iswimi.go
Comment (AldoFusterTurpin) [inline pkg/util/wimi/iswimi.go:14]: Following my first comment in the PR, if we want to have a richer domain and avoid an anemic domain model, we can move this function definition to pkg/api/openshiftcluster.go and have on line 29 something like:
```go
// IsWimi checks whether a cluster is a Workload Identity cluster or Service Principal cluster
func (oc *OpenShiftCluster) IsWimi() bool {
	if oc.Properties.PlatformWorkloadIdentityProfile != nil && oc.Properties.ServicePrincipalProfile == nil {
		return true
	}
	return false
}
```
which moves the business logic closer to where the data lives in 🙂

File: pkg/util/wimi/iswimi_test.go
Comment (AldoFusterTurpin) [inline pkg/util/wimi/iswimi_test.go:1]: Nice to see some unit tests 🙂

File: pkg/util/wimi/iswimi.go
Comment (AldoFusterTurpin) [reply pkg/util/wimi/iswimi.go:14]: One last minor comment about it, we can simplify the function if we change it to one single return where the true condition should be met 
```go
// IsWimi checks whether a cluster is a Workload Identity cluster or Service Principal cluster
func (oc *OpenShiftCluster) IsWimi() bool {
	return oc.Properties.PlatformWorkloadIdentityProfile != nil && oc.Properties.ServicePrincipalProfile == nil
}
```
Just mentioning it in case you find it easier to read.

**Review by AldoFusterTurpin — APPROVED** [review]
@slawande2 Thank you for this. LGTM! I just left an optional suggestion about moving the business logic closer to the struct where the data lives in to avoid having an anemic domain. No blocker at all, just a FYI 🙂

**Comment by cadenmarchese** [comment]
@AldoFusterTurpin thanks a lot for the suggestion! We are going to implement it in a new PR.

**Comment by AldoFusterTurpin** [comment]
> @AldoFusterTurpin thanks a lot for the suggestion! We are going to implement it in a new PR.

Great! I see that PR is now open: https://github.com/Azure/ARO-RP/pull/3683

Thanks!

**Pull request merged**
